### ISSUE TITLE
Custom relative path assetPrefix breaks HMR

### ISSUE THREAD

**Issue by fabb** [vercel/next.js#18080]
# Bug report

## Describe the bug

We have several next.js applications on the same domain, so we want to prefix `_next` requests. This works both for development and production:

```
module.exports = {
  assetPrefix: '/docs',
  rewrites() {
    return [
      { source: '/docs/_next/:path*', destination: '/_next/:path*' }
    ]
  }
}
```

The only thing that does not work are `webpack-hmr` calls. The calls will also be prefixed, but next.js returns a 404. Seems like the `rewrites` feature does not consider HMR calls.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Add the configuration mentioned above.
2. Start local development
3. See 404s for `webpack-hmr` calls in the browser network tab

## Expected behavior

The `rewrites` feature should also rewrite HMR requests. In fact, doing the rewrites in a custom express server fixes HMR, so it should be possible.

## System information

- OS: macOS Catalina
- Browser: Chrome 86
- Version of Next.js: 9.5.5
- Version of Node.js: 10.18.4

## Additional context

As mentioned here: https://github.com/vercel/next.js/issues/5602#issuecomment-708469829

**Comment by adam-foster** [comment]
Hi @fabb I experienced the same issue. 

As a work around, you can add the `assetPrefix` / `rewrites` config for non development phases only. I.e:

```
const { PHASE_DEVELOPMENT_SERVER } = require('next/constants')

module.exports = (phase, { defaultConfig }) => {
  if (phase !== PHASE_DEVELOPMENT_SERVER) {
    return {
      assetPrefix: '/docs',
      rewrites() {
        return [
           { source: '/docs/_next/:path*', destination: '/_next/:path*' }
        ]
      }
    }
  }
  return {
    /* development only config options here */
  }
}
```

**Comment by fabb** [comment]
> As a work around, you can add the assetPrefix / rewrites config for non development phases only.

In my case I cannot, because I have several next.js applications that I want to be able to run simultaneously in dev mode on localhost.

**Comment by meikidd** [comment]
I am getting same error in v10.2.3, curious any update on this issue?

**Comment by crutch12** [comment]
Bump

**Comment by crutch12** [comment]
Suddenly I found a fix!

Just add [redirects](https://nextjs.org/docs/api-reference/next.config.js/redirects):

```js
async redirects() {
  return [
    // @NOTE: rewrites breakes webpack-hmr, so we fix it here. See: https://github.com/vercel/next.js/issues/18080
    {
      source: '/:any*/_next/webpack-hmr:path*',
      destination: '/_next/webpack-hmr:path*',
      permanent: false,
    },
  ];
},
```

**Comment by fabb** [comment]
fixed in #30632

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
